falcon/routing/compiled.py

```diff
@@ -206,7 +241,19 @@ def insert(nodes, path_index=0):
                 insert(new_node.children, path_index + 1)
 
         insert(self._roots)
-        self._find = self._compile()
+
+        if kwargs.get('compile', False):
+            self._find = self._compile()
+        else:
+            if self._options.verify_route_on_add:
+                if self._verify_route_router is None:
+                    self._verify_route_router = CompiledRouter(_options=self._options)
+                kwargs['compile'] = True
+                self._verify_route_router._roots.clear()
```

Comment:
Why does the tree need to be reset each time?

Reply:
nevermind, I see that it is to speed things up since we are compiling. Would it be worth modifying the router class to do verification without compilation? Some errors will not be detected unless the entire tree is present (e.g., route conflicts).

Reply:
if we do not clear the roots we end up with the same behaviour as now, ie every time we will compile all the routes.

> Some errors will not be detected unless the entire tree is present (e.g., route conflicts).

I think insert will raise in this case here https://github.com/falconry/falcon/blob/c3e91875f1030c68d17efa27cf8df3d18f337761/falcon/routing/compiled.py#L186
and we are calling

I think that the route compile will raise if it does not find the converter for an id. I'm not sure if there are other cases.
Do you have an example of what we are missing not compiling all the routes?

>  Would it be worth modifying the router class to do verification without compilation?

Yes, I think we could do it, but I'm not sure if it would be worth it. The current times are excellent with the compilation step of the single route, and an user could also decide to remove it once he is sure that the routes work

Reply:
> I think that the route compile will raise if it does not find the converter for an id. I'm not sure if there are other cases.

I've had another look at it, and it seems that this is checked by the add_route function here
https://github.com/falconry/falcon/blob/c3e91875f1030c68d17efa27cf8df3d18f337761/falcon/routing/compiled.py#L169

So I'm not sure what we gain by compiling the routes one by one as It's done now by the `_verify_route_router`.
Maybe we could just omit it?

Reply:
I've tried to set the `verify_route_on_add` to false by default and the only failure in the tests are the one raised when the converter are instantiated, ie these ones

https://github.com/falconry/falcon/blob/c3e91875f1030c68d17efa27cf8df3d18f337761/tests/test_default_router.py#L454-L465

and I think that is because we are only instantiating the converter when compiling.

If we try to instantiate them in `_validate_template_segment`, just to test that it works and then throw the instance away, I think we could just omit the compile step.